rust/noosphere-core/src/data/body_chunk.rs

```diff
@@ -70,4 +76,315 @@ impl BodyChunkIpld {
+        }
+    }
+
+    /// Push a new chunk to be encoded.
+    pub async fn push(&mut self, chunk: Vec<u8>) -> Result<()> {
+        self.chunks.push(chunk).await?;
+        Ok(())
+    }
+
+    /// Consumes this [BodyChunkEncoder], encoding into the store, returning
+    /// the root [Cid] upon success.
+    pub async fn encode(mut self) -> Result<Cid> {
+        let mut next_chunk_cid = None;
+        let chunk_stream = self.chunks.as_stream_rev().await;
+        tokio::pin!(chunk_stream);
+
+        while let Some(chunk) = chunk_stream.try_next().await? {
+            next_chunk_cid = Some(
+                self.store
+                    .save::<DagCborCodec, _>(&BodyChunkIpld {
+                        bytes: Vec::from(chunk),
+                        next: next_chunk_cid,
+                    })
+                    .await?,
+            );
+        }
+        Ok(next_chunk_cid.unwrap())
+    }
+}
+
+/// Flexible vector byte-storage that stores items in memory, falling back to storing on disk
+/// if `memory_limit` is reached. Items can then be drained via a stream out of the structure.
+struct ChunkStorage {
+    chunks: ChunkStorageImpl,
+    memory_limit: u64,
+}
+
+impl ChunkStorage {
+    /// Creates a new [ChunkStorage] that stores data in memory
+    /// until `memory_limit` is reached.
+    pub fn new(memory_limit: u64) -> Self {
+        ChunkStorage {
+            chunks: ChunkStorageImpl::new_with_memory(),
+            memory_limit,
+        }
+    }
+
+    /// Push a new item.
+    pub async fn push(&mut self, chunk: Vec<u8>) -> Result<()> {
+        self.chunks.push(chunk).await?;
+
+        if matches!(self.chunks, ChunkStorageImpl::Memory { .. })
+            && self.chunks.byte_length() > &self.memory_limit
+        {
+            self.upgrade_storage().await?;
+        }
+
+        Ok(())
+    }
+
+    /// Drain all items as a stream.
+    #[allow(unused)]
+    pub async fn as_stream(self) -> impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Unpin {
+        self.chunks.as_stream(false).await
+    }
+
+    /// Drain all items as a stream in reverse.
+    pub async fn as_stream_rev(
+        self,
+    ) -> impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Unpin {
+        self.chunks.as_stream(true).await
+    }
+
+    async fn upgrade_storage(&mut self) -> Result<()> {
+        let mut storage = ChunkStorageImpl::new_with_disk()?;
+        std::mem::swap(&mut self.chunks, &mut storage);
+        let previous_stream = storage.as_stream(false).await;
+        tokio::pin!(previous_stream);
+        while let Some(chunk) = previous_stream.try_next().await? {
+            self.chunks.push(chunk).await?;
+        }
+        Ok(())
+    }
+}
+
+enum ChunkStorageImpl {
+    Memory {
+        chunks: Vec<Vec<u8>>,
+        chunk_count: usize,
+        byte_length: u64,
+    },
+    Disk {
+        temp_dir: tempfile::TempDir,
+        chunk_count: usize,
+        byte_length: u64,
+    },
+}
+
+impl ChunkStorageImpl {
```

Comment:
Please avoid abbreviations e.g., `Impl`

Reply:
Open to suggestions on `ChunkStorage` (:woozy_face: ) as well!